kube-core/src/labels.rs

```diff
@@ -0,0 +1,552 @@
+    /// use k8s_openapi::apimachinery::pkg::apis::meta::v1::LabelSelector;
+    /// use kube_core::SelectorExt;
+    /// use kube_core::Expression;
+    /// let matches = Pod::default().selector_matches(&LabelSelector::default());
+    /// assert!(matches);
+    /// let matches = Pod::default().selector_matches(&Expression::Exists("foo".into()));
+    /// assert!(!matches);
+    /// ```
+    fn selector_matches(&self, selector: &impl Matcher) -> bool {
+        selector.matches(self.selector_map())
+    }
+}
+
+impl<R: ResourceExt> SelectorExt for R {
+    fn selector_map(&self) -> &BTreeMap<String, String> {
+        self.labels()
+    }
+}
+
+/// Matcher trait for implementing alternalive Selectors
+pub trait Matcher {
+    // Perform a match check on the resource labels
+    fn matches(&self, labels: &BTreeMap<String, String>) -> bool;
+}
+
+/// A selector expression with existing operations
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+pub enum Expression {
+    /// Key exists and in set:
+    ///
+    /// ```
+    /// use kube_core::Expression;
+    ///
+    /// let exp = Expression::In("foo".into(), ["bar".into(), "baz".into()].into()).to_string();
+    /// assert_eq!(exp, "foo in (bar,baz)");
+    /// let exp = Expression::In("foo".into(), vec!["bar".into(), "baz".into()].into_iter().collect()).to_string();
+    /// assert_eq!(exp, "foo in (bar,baz)");
+    /// ```
+    In(String, BTreeSet<String>),
+
+    /// Key does not exists or not in set:
+    ///
+    /// ```
+    /// use kube_core::Expression;
+    ///
+    /// let exp = Expression::NotIn("foo".into(), ["bar".into(), "baz".into()].into()).to_string();
+    /// assert_eq!(exp, "foo notin (bar,baz)");
+    /// let exp = Expression::NotIn("foo".into(), vec!["bar".into(), "baz".into()].into_iter().collect()).to_string();
+    /// assert_eq!(exp, "foo notin (bar,baz)");
+    /// ```
+    NotIn(String, BTreeSet<String>),
+
+    /// Key exists and is equal:
+    ///
+    /// ```
+    /// use kube_core::Expression;
+    ///
+    /// let exp = Expression::Equal("foo".into(), "bar".into()).to_string();
+    /// assert_eq!(exp, "foo=bar")
+    /// ```
+    Equal(String, String),
+
+    /// Key does not exists or is not equal:
+    ///
+    /// ```
+    /// use kube_core::Expression;
+    ///
+    /// let exp = Expression::NotEqual("foo".into(), "bar".into()).to_string();
+    /// assert_eq!(exp, "foo!=bar")
+    /// ```
+    NotEqual(String, String),
+
+    /// Key exists:
+    ///
+    /// ```
+    /// use kube_core::Expression;
+    ///
+    /// let exp = Expression::Exists("foo".into()).to_string();
+    /// assert_eq!(exp, "foo")
+    /// ```
+    Exists(String),
+
+    /// Key does not exist:
+    ///
+    /// ```
+    /// use kube_core::Expression;
+    ///
+    /// let exp = Expression::DoesNotExist("foo".into()).to_string();
+    /// assert_eq!(exp, "!foo")
+    /// ```
+    DoesNotExist(String),
+
+    /// Invalid combination. Always evaluates to false.
```

Comment:
Is the "always evaluates to false" behavior borrowed from anywhere else or just an arbitrary choice?

Reply:
An alternative to having an error scenario, yes, an arbitrary choice. I removed it, as having an error is more convenient.